### PR TITLE
CVSL-1616 Adding subject access request endpoint

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/AuditEvent.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    return "Audit(id=$id, timestamp='$timestamp', user=$user, action=$action, details= $details)"
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (other !is AuditEvent) return false
+
+    if (id != other.id) return false
+
+    return true
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/AuditEvent.kt:41]: Minor, could 39-41 be replaced with `return id == other.id`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/AuditEventRepository.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor
+import org.springframework.data.jpa.repository.Query
+import org.springframework.stereotype.Repository
+
+@Repository
+interface AuditEventRepository : JpaRepository<AuditEvent, Long>, JpaSpecificationExecutor<AuditEvent> {
+
+  @Query(
+    nativeQuery = true,
+    value = "select id, timestamp, user, action, details from audit where details ->> 'bookingId' in :bookingIds order by timestamp desc;",
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/AuditEventRepository.kt:13]: Question for learning, not a blocker

Given the structure of `details` is a map, is this saying `where the booking ID in the details map is in the list of booking IDs provided`? If so, very cool!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/subjectaccessrequests/SubjectAccessRequestController.kt [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        content = [
+          Content(
+            mediaType = "application/json",
+            schema = Schema(implementation = SarContent::class),
+          ),
+        ],
+      ),
+      ApiResponse(
+        responseCode = "204",
+        description = "Records for this prn was not found.",
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/subjectaccessrequests/SubjectAccessRequestController.kt:69]: Does this not need a content block to return the `NO_CONTENT` error on line 34?

**Review by SebNegahbanMoJ — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/subjectaccessrequests/SubjectAccessRequestController.kt (hunk 3 above)
Comment (andrewrlee) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/subjectaccessrequests/SubjectAccessRequestController.kt:69]: yes, it does - I'll fix in a follow up

**Pull request merged**